examples/transactions.c

```diff
@@ -321,112 +293,125 @@ static void commit_transaction_and_start_new (struct state *state) {
-        error = rd_kafka_commit_transaction(state->producer, -1);
+        error = rd_kafka_commit_transaction(producer, -1);
         if (error) {
                 if (rd_kafka_error_txn_requires_abort(error)) {
                         fprintf(stderr,
                                 "WARNING: Failed to commit transaction: "
                                 "%s: %s: aborting transaction\n",
                                 rd_kafka_error_name(error),
                                 rd_kafka_error_string(error));
-                        abort_transaction_and_rewind(state);
                         rd_kafka_error_destroy(error);
-                        return;
+
+                        /* Abort transaction */
+                        error = rd_kafka_abort_transaction(consumer, -1);
+                        if (error)
+                                fatal_error("Failed to abort transaction",
+                                            error);
+                        return 0;
                 } else {
                         fatal_error("Failed to commit transaction", error);
                 }
         }
 
+        return 1;
+}
+
+/**
+ * @brief Commit the current transaction and start a new transaction.
+ */
+static void commit_transaction_and_start_new (rd_kafka_t *consumer,
+                                              rd_kafka_t *producer) {
+        rd_kafka_error_t *error;
+
+        /* Commit transaction.
+         * If commit failed the transaction is aborted and we need
+         * to rewind the consumer to the last committed offsets. */
+        if (!commit_transaction(consumer, producer))
+                rewind_consumer(consumer);
+
         /* Begin new transaction */
-        error = rd_kafka_begin_transaction(state->producer);
+        error = rd_kafka_begin_transaction(producer);
         if (error)
                 fatal_error("Failed to begin new transaction", error);
 }
 
 /**
  * @brief The rebalance will be triggered (from rd_kafka_consumer_poll())
  *        when the consumer's partition assignment is assigned or revoked.
- *
- * Prior to KIP-447 being supported there must be one transactional output
- * producer for each consumed input partition, so we create and destroy
- * these producer's from this callback.
  */
 static void
-consumer_group_rebalance_cb (rd_kafka_t *rk,
+consumer_group_rebalance_cb (rd_kafka_t *consumer,
                              rd_kafka_resp_err_t err,
                              rd_kafka_topic_partition_list_t *partitions,
                              void *opaque) {
-        int i;
+        rd_kafka_t *producer = (rd_kafka_t *)opaque;
+        rd_kafka_error_t *error;
 
-        if (!strcmp(rd_kafka_rebalance_protocol(rk), "COOPERATIVE"))
-                fatal("This example has not yet been modified to work with "
-                      "cooperative incremental rebalancing "
+        if (strcmp(rd_kafka_rebalance_protocol(consumer), "COOPERATIVE"))
```

Comment:
the incremental assign will fail in this isn't the case - i don't think checking this is necessary. i only added the `rd_kafka_rebalance_protocol` method because it was required by the C# binding. generally, I don't think users should use it.

Reply:
:+1: